File: igor/igor-web/src/main/groovy/com/netflix/spinnaker/igor/admin/AdminController.java
/*
 * Copyright 2018 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.spinnaker.igor.admin;

import static java.lang.String.format;

import com.netflix.spinnaker.igor.polling.CommonPollingMonitor;
import com.netflix.spinnaker.igor.polling.PollingMonitor;
import com.netflix.spinnaker.kork.web.exceptions.NotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/admin")
public class AdminController {

  private final Logger log = LoggerFactory.getLogger(getClass());

  private final List<CommonPollingMonitor> pollingMonitors;

  @Autowired
  public AdminController(Optional<List<CommonPollingMonitor>> pollingMonitors) {
    this.pollingMonitors = pollingMonitors.orElseGet(ArrayList::new);
  }

  /**
   * Silently fast-forwards a poller. Fast-forwarding means that all pending cache state will be
   * polled and saved, but will not send Echo notifications.
   *
   * <p>By default all partitions (ex: masters) will be fast-forwarded, however specific partition
   * names should be used whenever possible.
   *
   * @param monitorName The polling monitor name (ex: "DockerMonitor")
   * @param partition The partition name, if not provided, method will re-index the entire monitor
   */
  @RequestMapping(value = "/pollers/fastforward/{monitorName}", method = RequestMethod.POST)
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void fastForward(
      @PathVariable String monitorName, @RequestParam(required = false) String partition) {
    CommonPollingMonitor pollingMonitor =
        pollingMonitors.stream()
            .filter(it -> it.getName().equals(monitorName))
            .findFirst()
            .orElseThrow(
                () ->
                    new NotFoundException(
                        format(
                            "PollingMonitor %s was not found, available monitors are: %s",
                            monitorName,
                            pollingMonitors.stream()
                                .map(PollingMonitor::getName)
                                .collect(Collectors.toList()))));

    log.warn("Re-indexing {}:{}", monitorName, (partition == null) ? "ALL" : partition);
    if (partition == null) {
      pollingMonitor.poll(false);
    } else {
      pollingMonitor.pollSingle(pollingMonitor.getPollContext(partition).fastForward());
    }
  }
}

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/AdminController.kt
package com.netflix.spinnaker.keel.rest

import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
import com.netflix.spinnaker.keel.persistence.NoSuchApplication
import com.netflix.spinnaker.keel.persistence.NoSuchResourceException
import com.netflix.spinnaker.keel.persistence.ResourceRepository
import org.slf4j.LoggerFactory.getLogger
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.NO_CONTENT
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.ExceptionHandler

@RestController
@RequestMapping(path = ["/admin"])
class AdminController(
  private val deliveryConfigRepository: DeliveryConfigRepository,
  private val resourceRepository: ResourceRepository
) {
  private val log by lazy { getLogger(javaClass) }

  @DeleteMapping(
      path = ["/applications/{application}"]
  )
  @ResponseStatus(NO_CONTENT)
  fun deleteApplicationData(
    @PathVariable("application") application: String
  ) {
    log.debug("Deleting all data for application: $application")
    deleteApplication(application)
  }

  @ExceptionHandler(NoSuchResourceException::class)
  @ResponseStatus(HttpStatus.NOT_FOUND)
  fun onNotFound(e: NoSuchResourceException) {
    log.error(e.message)
  }

  fun deleteApplication(application: String) {
    val resources = resourceRepository.deleteByApplication(application)
    val deliveryConfigs = deliveryConfigRepository.deleteByApplication(application)
    if (resources == 0 && deliveryConfigs == 0)
      throw NoSuchApplication(application)
  }
}
